File: cave/com.raytheon.viz.gfe/src/com/raytheon/viz/gfe/GfeClient.java
/**
 * This software was developed and / or modified by Raytheon Company,
 * pursuant to Contract DG133W-05-CQ-1067 with the US Government.
 * 
 * U.S. EXPORT CONTROLLED TECHNICAL DATA
 * This software product contains export-restricted data whose
 * export/transfer/disclosure is restricted by U.S. law. Dissemination
 * to non-U.S. persons whether in the United States or abroad requires
 * an export license or other authorization.
 * 
 * Contractor Name:        Raytheon Company
 * Contractor Address:     6825 Pine Street, Suite 340
 *                         Mail Stop B8
 *                         Omaha, NE 68106
 *                         402.291.0100
 * 
 * See the AWIPS II Master Rights File ("Master Rights File.pdf") for
 * further licensing information.
 **/
package com.raytheon.viz.gfe;

import java.io.File;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashSet;
import java.util.Timer;
import java.util.TimerTask;

import jep.Jep;
import jep.JepConfig;
import jep.JepException;
import jep.NamingConventionClassEnquirer;

import org.eclipse.core.runtime.FileLocator;
import org.eclipse.core.runtime.Path;
import org.eclipse.core.runtime.Platform;

import com.raytheon.uf.common.dataplugin.gfe.python.GfePyIncludeUtil;
import com.raytheon.uf.common.python.PyUtil;
import com.raytheon.uf.common.python.PythonIncludePathUtil;
import com.raytheon.uf.common.status.IUFStatusHandler;
import com.raytheon.uf.common.status.UFStatus;
import com.raytheon.uf.common.status.UFStatus.Priority;
import com.raytheon.uf.common.time.util.TimeUtil;
import com.raytheon.uf.common.util.FileUtil;
import com.raytheon.uf.viz.core.VizApp;
import com.raytheon.viz.gfe.python.GfeCavePyIncludeUtil;
import com.raytheon.viz.ui.personalities.awips.AbstractAWIPSComponent;

/**
 * GFE Client application component
 * 
 * <pre>
 * 
 * SOFTWARE HISTORY
 * Date         Ticket#    Engineer    Description
 * ------------ ---------- ----------- --------------------------
 * Jun 25, 2010            mschenke     Initial creation
 * Aug 20, 2012  #1081     dgilling     Don't pass -server and -site args
 *                                      to python script.
 * Sep 11, 2013  #2033     dgilling     Update path to utilityDir and pyVizDir,
 *                                      now that they're no longer in 
 *                                      localization store.
 * Dec 04, 2013  #2588     dgilling     Add thread to force shutdown.
 * Mar 25, 2014  #2963     randerso     Removed obsolete python_include support 
 *                                      which was adding an empty string into the 
 *                                      python path causing python to look in user's
 *                                      current default directory for modules.
 * Aug 22, 2014  3500      bclement     override postStartupActions()
 * Aug 29, 2014  3500      bclement     removed override of postStartupActions() 
 *                                      since ProcedureXMLManager startup was moved to the CAVE subclass
 * Apr 26, 2015  4259      njensen      Updated for new JEP API
 * May 20, 2015  4509      njensen      Added time and dataaccess to include path
 * Apr 28, 2016  5236      njensen      Use Jep redirectOutput for python prints
 * Jul 19, 2017  ----      mjames@ucar  Add common python include path.
 * 
 * </pre>
 * 
 * @author mschenke
 */

public class GfeClient extends AbstractAWIPSComponent {

    private static final transient IUFStatusHandler statusHandler = UFStatus
            .getHandler(GfeClient.class);

    @Override
    protected void startInternal(String componentName) throws Exception {
        long t0 = System.currentTimeMillis();

        String[] args = Platform.getApplicationArgs();
        int gfeClientArgStartIndex = -1;

        for (int i = 0; i < args.length; ++i) {
            String arg = args[i];
            if (componentName.equals(arg)) {
                gfeClientArgStartIndex = i + 1;
                break;
            }
        }

        if ((gfeClientArgStartIndex == -1)
                || (gfeClientArgStartIndex == args.length)) {
            System.err.println("No python script specified to run - exiting");
            return;
        }

        System.out.println("Running script: " + args[gfeClientArgStartIndex]);

        String pyVizDir = new File(FileLocator.resolve(
                FileLocator.find(Activator.getDefault().getBundle(), new Path(
                        FileUtil.join("python", "pyViz")), null)).getPath())
                .getPath();

        String utilityDir = new File(FileLocator.resolve(
                FileLocator.find(Activator.getDefault().getBundle(), new Path(
                        FileUtil.join("python", "utility")), null)).getPath())
                .getPath();

        boolean includeUser = (!VizApp.getWsId().getUserName().equals("SITE"));

        String includePath = PyUtil.buildJepIncludePath(true, utilityDir,
                PythonIncludePathUtil.getCommonPythonIncludePath(),
                PythonIncludePathUtil.getCommonPythonIncludePath("time",
                        "dataaccess"), GfePyIncludeUtil
                        .getCommonGfeIncludePath(), GfePyIncludeUtil
                        .getConfigIncludePath(includeUser), pyVizDir,
                GfePyIncludeUtil.getUtilitiesIncludePath(includeUser),
                GfePyIncludeUtil.getIToolIncludePath(), GfePyIncludeUtil
                        .getVtecIncludePath(), GfePyIncludeUtil
                        .getHeadlineIncludePath(), GfeCavePyIncludeUtil
                        .getAutotestIncludePath(), GfePyIncludeUtil
                        .getTextUtilitiesIncludePath(includeUser),
                GfePyIncludeUtil.getTextProductsIncludePath(includeUser),
                GfePyIncludeUtil.getCombinationsIncludePath(includeUser),
                GfeCavePyIncludeUtil.getTestsIncludePath(), GfePyIncludeUtil
                        .getProceduresIncludePath(includeUser));

        Jep jep = null;
        try {
            JepConfig config = new JepConfig().setInteractive(false)
                    .setIncludePath(includePath)
                    .setClassLoader(GfeClient.class.getClassLoader())
                    .setClassEnquirer(new NamingConventionClassEnquirer())
                    .setRedirectOutputStreams(true);
            jep = new Jep(config);
            jep.eval("import sys");
            jep.eval("sys.argv = []");
            boolean skipNextArg = false;
            Collection<String> ignoredParams = getIgnoredParameters();
            for (int i = gfeClientArgStartIndex; i < args.length; i++) {
                if (ignoredParams.contains(args[i])) {
                    skipNextArg = true;
                } else if (skipNextArg) {
                    skipNextArg = false;
                } else {
                    jep.eval("sys.argv.append('"
                            + args[i].replaceAll("'", "\\\\'") + "')");
                }
            }
            jep.runScript(args[gfeClientArgStartIndex]);
            // jep.eval("main()");
        } catch (JepException e) {
            statusHandler.handle(Priority.ERROR, e.getLocalizedMessage(), e);
        }
        if (jep != null) {
            jep.close();
        }
        long t1 = System.currentTimeMillis();
        System.out.println("Entire execution to run python script: "
                + (t1 - t0));

        // operationally, we've found situations where gfeclient jobs seem to
        // hang around running even though all non-daemon threads have completed
        // their work. So, in attempt to prevent those cases from hanging around
        // as "zombie" processes let's set a timer to kill the JVM if things
        // haven't exited by themselves.
        Timer shutdownTimer = new Timer("gfe-client-shutdown", true);
        TimerTask shutdownTask = new TimerTask() {

            @Override
            public void run() {
                statusHandler
                        .warn("GFEClient should have already exited, but it hasn't. Manually exiting.");
                System.exit(0);
            }
        };
        shutdownTimer.schedule(shutdownTask, 45 * TimeUtil.MILLIS_PER_SECOND);
    }

    @Override
    protected int getRuntimeModes() {
        return (NON_UI | ALERT_VIZ);
    }

    private Collection<String> getIgnoredParameters() {
        return new HashSet<>(
                Arrays.asList("-site", "-server", "-mode", "-time"));
    }

}

File: cave/com.raytheon.viz.gfe/GFESuite/src/gfeClient/gfeClient.py
#!/awips2/python/bin/python3
##
# This software was developed and / or modified by Raytheon Company,
# pursuant to Contract DG133W-05-CQ-1067 with the US Government.
#
# U.S. EXPORT CONTROLLED TECHNICAL DATA
# This software product contains export-restricted data whose
# export/transfer/disclosure is restricted by U.S. law. Dissemination
# to non-U.S. persons whether in the United States or abroad requires
# an export license or other authorization.
#
# Contractor Name:        Raytheon Company
# Contractor Address:     6825 Pine Street, Suite 340
#                         Mail Stop B8
#                         Omaha, NE 68106
#                         402.291.0100
#
# See the AWIPS II Master Rights File ("Master Rights File.pdf") for
# further licensing information.
##
# -----------------------------------------------------------------------------
#
# SOFTWARE HISTORY
#
# Date          Ticket#  Engineer  Description
# ------------- -------- --------- ---------------------------------------------
# Jan 24, 2017  6092     randerso  Initial Creation
#
##

import os
import sys
import time
import argparse

from dynamicserialize.dstypes.com.raytheon.uf.common.dataplugin.gfe.request.GfeClientRequest import GfeClientRequest
from dynamicserialize.dstypes.java.util import Date
from ufpy import ThriftClient
from ufpy import UsageArgumentParser
from ufpy.UsageArgumentParser import StoreTimeAction
from ufpy.UsageArgumentParser import TIME_FORMAT

def validateArgs(args=None):

    parser = UsageArgumentParser.UsageArgumentParser(conflict_handler="resolve",
                                                     prog='gfeClient', add_help=False)
    parser.add_argument("script", action="store",
                      help=argparse.SUPPRESS,
                      metavar="script")
    parser.add_argument("-h", "--host", action="store", dest="host",
                      help="host name of edex request server",
                      default=str(os.getenv("DEFAULT_HOST", "localhost")),
                      metavar="hostname")
    parser.add_argument("-p", "--port", action="store", type=int, dest="port",
                      help="port number of edex request server",
                      default=int(os.getenv("DEFAULT_PORT", "9581")),
                      metavar="port")

    ############################################################################
    # -site is used for backward compatibility, --site is preferred
    # long names with single dash are non-standard in Unix/Linux
    ############################################################################
    parser.add_argument("--site", "-site", action="store", dest="site", required=True,
                      help="site ID",
                      metavar="site")
    parser.add_argument("-c", "--config", action="store", dest="configFile", required=False,
                      default="gfeConfig",
                      help="GFE config file -- default gfeConfig",
                      metavar="configFile")
    parser.add_argument("-u", action="store", dest="userName", required=False,
                        help="user name -- default SITE",
                        default="SITE",
                        metavar="userName")
    parser.add_argument("-z", "--drt", action=StoreTimeAction, dest="drt", required=False,
                      help="displaced real time -- format YYYYMMDD_hhmm",
                      metavar="drt")

    ############################################################################
    # adding this arguments so -s is not recognized as -site in other scripts
    # -s is not used by this script
    ############################################################################
    parser.add_argument("-s", action="store", dest="startTime", required=False,
                        help=argparse.SUPPRESS)

    args, scriptArgs = parser.parse_known_args(args)
    return parser, args, scriptArgs

def main(args):

    # if no args other than script add --help so usage is displayed
    if len(args) < 2:
        args.extend(["--help"])

    # if --help in args add dummy --site arg so we can display
    # full script usage, not just the gfeClient.py usage
    if "--help" in args:
        args.extend(["--site", "XXX"])

    parser, gfeClientArgs, scriptArgs = validateArgs(args)

    # add config and user option to scriptArgs
    scriptArgs.extend(["-c", gfeClientArgs.configFile, "-u", gfeClientArgs.userName])

    # add drt option if specified
    if gfeClientArgs.drt:
        timeString = time.strftime(TIME_FORMAT, gfeClientArgs.drt)
        scriptArgs.extend(["-z", timeString])

    # add startTime option if specified
    if gfeClientArgs.startTime:
        scriptArgs.extend(["-s", gfeClientArgs.startTime])

    # shutdown isn't a real script and has no gfeClientArgs to validate
    if gfeClientArgs.script.lower() != "shutdown":

        # call the validateArgs() method in the target script
        scriptGlobals = {}
        scriptLocals = {}
        exec(compile(open(gfeClientArgs.script, "rb").read(), gfeClientArgs.script, 'exec'), scriptGlobals, scriptLocals)
        scriptLocals["validateArgs"](args, [parser])

    elif "--help" in args:
        # Don't do shutdown if --help specified
        # this is only for ifpIMAGE since it's calling shutdown until
        # PngWriter can be fixed to run more than once in a session
        sys.exit(0)

    request = GfeClientRequest(gfeClientArgs.script, gfeClientArgs.site,
                               gfeClientArgs.configFile, gfeClientArgs.userName,
                               scriptArgs)
    if gfeClientArgs.drt:
        import calendar

        timeInMillis = calendar.timegm(gfeClientArgs.drt) * 1000
        request.setTime(Date(timeInMillis))

    thriftClient = ThriftClient.ThriftClient(gfeClientArgs.host, gfeClientArgs.port, "/services")
    thriftClient.sendRequest(request)

if __name__ == '__main__':
    sys.exit(main(sys.argv[1:]))
